### ISSUE TITLE
Exception handling

### ISSUE THREAD

**Issue by lassik** [udem-dlteam/ribbit#10]
Exception handling (`raise` / `guard`) is mandatory for useful Scheme programming. What's the plan for supporting it? I assume it can be implemented using continuations; is that usable enough?

**Comment by feeley** [comment]
Yes exception handling can be implemented on top of continuations.  Ribbit currently supports `call/cc` so `raise`/`guard` should be easy to add.  The next step would be to add macros to Ribbit so that these extensions can be put in a library rather than the compiler proper.

The main design issue is how to implement these extensions in a modular way so that it doesn't bloat the system if you don't make use of these features.  The current system including a REPL fits in 2 KB of RVM code on top of a 2 KB RVM implementation.  I would like to explore how much of R7RS can be implemented in twice that (which is really 3 times the RVM code because the RVM implementation will probably not have to change).

An issue related to exception handling is support for debugging.  Raising an exception is simple, but offering a useful debugging environment is not simple (what information to put in error messages, ability to inspect the continuation and move around, line numbers, etc).  These features tend to bloat the system.  Ideally there should be a way to turn off support for debugging.

**Comment by raviqqe** [comment]
> The next step would be to add macros to Ribbit so that these extensions can be put in a library rather than the compiler proper.

In this case, do you think the macros should be implemented entirely by RVM primitives like the current implementation of REPL, just the AOT compiler, or a mixture of those?

BTW, thank you for this exciting project! I have no idea why it's so fast while the bytecode representation in ribs looks bad for data access locality. 🤯
